### PR TITLE
K8SPSMDB-282 integrate pmm into PSMDBO

### PR DESCRIPTION
**Pull request by hors**
[K8SPXC-497](https://jira.percona.com/browse/K8SPXC-497)

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | failed |
| limits | passed |
| scaling | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| monitoring-2-0 | failed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| monitoring | failed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| arbiter | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | failed |
| scaling | passed |
| self-healing | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| security-context | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | failed |
| users | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| init-deploy | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| init-deploy | passed |
| self-healing | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| monitoring-2-0 | failed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| limits | passed |
| monitoring | failed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| monitoring-2-0 | failed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| arbiter | passed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by delgod — APPROVED** [review]

File: pkg/psmdb/pmm.go [hunk 1]
```diff
@@ -142,5 +143,131 @@ func PMMContainer(spec api.PMMSpec, secrets string, customLogin bool, clusterNam
...
+
+func PMMAgentScript() []corev1.EnvVar {
+	pmmServerArgs := " --skip-connection-check --metrics-mode=push "
+	pmmServerArgs = pmmServerArgs + " --username=$(DB_USER) --password=$(DB_PASSWORD) --cluster=$(DB_CLUSTER) "
+	pmmServerArgs = pmmServerArgs + "--service-name=$(PMM_AGENT_SETUP_NODE_NAME) --host=$(DB_HOST) --port=$(DB_PORT)"
+
+	return []corev1.EnvVar{
+		{
+			Name:  "PMM_AGENT_PRERUN_SCRIPT",
+			Value: "sleep 10; pmm-admin add $(DB_TYPE)" + pmmServerArgs + "; pmm-admin annotate reboot",
```
Comment (delgod) [inline pkg/psmdb/pmm.go:270]:
```suggestion
			Value: "sleep 10; pmm-admin add $(DB_TYPE)" + pmmServerArgs + "; pmm-admin annotate restart",
```

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | failed |
| upgrade | failed |
| limits | passed |
| scaling | passed |
| self-healing | passed |
| security-context | failed |
| monitoring | failed |
| upgrade-consistency | passed |
| demand-backup | failed |
| operator-self-healing | passed |
| monitoring-2-0 | failed |
| arbiter | passed |
| scheduled-backup | failed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

File: pkg/psmdb/pmm.go (hunk 1 above)
Comment (hors) [reply pkg/psmdb/pmm.go]: fixed

File: pkg/psmdb/pmm.go [hunk 2]
```diff
@@ -142,5 +143,131 @@ func PMMContainer(spec api.PMMSpec, secrets string, customLogin bool, clusterNam
...
+
+func PMMAgentScript() []corev1.EnvVar {
+	pmmServerArgs := " --skip-connection-check --metrics-mode=push "
+	pmmServerArgs = pmmServerArgs + " --username=$(DB_USER) --password=$(DB_PASSWORD) --cluster=$(DB_CLUSTER) "
+	pmmServerArgs = pmmServerArgs + "--service-name=$(PMM_AGENT_SETUP_NODE_NAME) --host=$(DB_HOST) --port=$(DB_PORT)"
+
+	return []corev1.EnvVar{
+		{
+			Name:  "PMM_AGENT_PRERUN_SCRIPT",
+			Value: "sleep 10; pmm-admin add $(DB_TYPE)" + pmmServerArgs + "; pmm-admin annotate restart",
```
Comment (delgod) [inline pkg/psmdb/pmm.go:270]: please use `pmm-admin status --wait` 
```suggestion
			Value: "pmm-admin status --wait; pmm-admin add $(DB_TYPE)" + pmmServerArgs + "; pmm-admin annotate restart",
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/psmdb/pmm.go (hunk 2 above)
Comment (hors) [reply pkg/psmdb/pmm.go]: yes, I know it. I have updated this PR

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | failed |
| upgrade | failed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| operator-self-healing | passed |
| security-context | failed |
| monitoring | failed |
| upgrade-consistency | passed |
| demand-backup | failed |
| monitoring-2-0 | failed |
| arbiter | passed |
| scheduled-backup | failed |
| service-per-pod | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | failed |

**Review by delgod — CHANGES_REQUESTED** [review]
it is needed to add param options for mongod and mongos

**Comment by hors** [comment]
> it is needed to add param options for mongod and mongos

yes, we need but firstly we need to add pmm client to mongos via a separate PR. And I have a separate task regarding param options. I will appreciate very much if you all stop adding additional requirements to this PR :) Let's  try to merge it.

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | failed |
| storage | passed |
| init-deploy | passed |
| upgrade | failed |
| self-healing | passed |
| limits | passed |
| scaling | passed |
| security-context | failed |
| monitoring | failed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| demand-backup | failed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| self-healing | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |

File: pkg/psmdb/pmm.go [hunk 3]
```diff
@@ -12,7 +13,7 @@ const (
 )
 
 // PMMContainer returns a pmm container from given spec
-func PMMContainer(spec api.PMMSpec, secrets string, customLogin bool, clusterName string, v120OrGreater bool) corev1.Container {
+func PMMContainer(spec api.PMMSpec, secrets string, customLogin bool, clusterName string, v120OrGreater bool, v170OrGreater bool) corev1.Container {
```
Comment (delgod) [inline pkg/psmdb/pmm.go:16]:
```suggestion
func PMMContainer(spec api.PMMSpec, secrets string, customLogin bool, clusterName string, v120OrGreater bool, v160OrGreater bool) corev1.Container {
```

**Review by delgod — CHANGES_REQUESTED** [review]

File: pkg/psmdb/pmm.go (hunk 3 above)
Comment (hors) [reply pkg/psmdb/pmm.go:16]: thank you!

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| init-deploy | passed |
| one-pod | failed |
| upgrade | failed |
| limits | passed |
| scaling | passed |
| self-healing | passed |
| security-context | failed |
| monitoring | failed |
| upgrade-consistency | passed |
| operator-self-healing | passed |
| demand-backup | failed |
| scheduled-backup | failed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| one-pod | passed |
| storage | passed |
| upgrade | passed |
| monitoring | failed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| operator-self-healing | passed |
| demand-backup-sharded | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| one-pod | passed |
| init-deploy | passed |
| storage | passed |
| upgrade | passed |
| monitoring | failed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| scaling | passed |
| demand-backup | passed |
| operator-self-healing | passed |
| security-context | passed |
| scheduled-backup | passed |
| demand-backup-sharded | passed |

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-443-0b7ba606

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| storage | passed |
| one-pod | passed |
| init-deploy | passed |
| upgrade | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| arbiter | passed |
| scaling | passed |
| demand-backup | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| demand-backup-sharded | passed |
| liveness | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Review by delgod — CHANGES_REQUESTED** [review]

File: Jenkinsfile [hunk 4]
```diff
@@ -264,7 +266,6 @@ pipeline {
                     steps {
                         CreateCluster('basic')
                         runTest('one-pod', 'basic')
-                        runTest('monitoring', 'basic')
```
[the remark this answers is not in the document]
Comment (hors) [reply Jenkinsfile:267]: It is test for PMM1. We need to support official PMM2 only. Why should we support unofficial pmm docker image? P.S. I have discussed it with @spron-in  and he confirmed that we need to support only official PMM2.

[the remark this answers is not in the document]
Comment (delgod) [reply Jenkinsfile:267]: can we remove it at all in such case?

[the remark this answers is not in the document]
Comment (hors) [reply Jenkinsfile:267]: yes, I will clean up the old code  in separate PR.

**Review by delgod — APPROVED** [review]

**Review by qJkee — APPROVED** [review]

**Comment by JNKPercona** [comment]
PSMDB operator docker - perconalab/percona-server-mongodb-operator:PR-443-7c14cd53

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| init-deploy | passed |
| storage | passed |
| one-pod | passed |
| upgrade | passed |
| monitoring-2-0 | passed |
| self-healing | passed |
| limits | passed |
| upgrade-consistency | passed |
| demand-backup | passed |
| scaling | passed |
| arbiter | passed |
| security-context | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| service-per-pod | passed |
| liveness | passed |
| demand-backup-sharded | passed |
| smart-update | passed |
| version-service | passed |
| users | passed |

**Pull request merged**